framework/auth/views.py

```diff
@@ -198,88 +198,155 @@ def forgot_password_post(auth, **kwargs):
+    if not logout and auth.logged_in:
+        data['status_code'] = http.FOUND
+        data['redirect_url'] = service_url
+        data['must_login_waring'] = False
+
+    # handle `claim_user_registered`
+    if logout:
+        if auth.logged_in:
+            data['status_code'] = 'auth_logout'
+            data['redirect_url'] = request.url
+        else:
+            data['must_login_warning'] = True
 
-        GET '/login/?campaign=prereg:
-            if user is logged in, redirect to prereg home page
-            else show sign up page and notify user to sign in, set next to prereg home page
+    return data
 
-        GET '/login/?next=next_url:
-            if user is logged in, redirect to next_url
-            else redirect to CAS login page with next_url as target service
+
+def validate_campaign(campaign, **kwargs):
     """
+    Non-view helper function that checks `campaign`.
 
-    campaign = request.args.get('campaign')
-    next_url = request.args.get('next')
-    log_out = request.args.get('logout')
-    must_login_warning = True
+    :param campaign: the campaign to check
+    :param kwargs: None
+    :return: True if valid, False otherwise
+    """
 
-    if not campaign and not next_url and not log_out:
-        if auth.logged_in:
-            return redirect(web_url_for('dashboard'))
-        return redirect(cas.get_login_url(web_url_for('dashboard', _absolute=True)))
+    return campaign and campaign in campaigns.CAMPAIGNS
 
-    if campaign:
-        next_url = campaigns.campaign_url_for(campaign)
 
-    if not next_url:
-        next_url = request.args.get('redirect_url')
-        must_login_warning = False
+def validate_service_url(service_url, **kwargs):
+    """
+    Non-view helper function that checks `service_url`. Only allow redirects which are relative root or full
+    domain. Disallows external redirects.
 
-    if next_url:
-        # Only allow redirects which are relative root or full domain, disallows external redirects.
-        if not (next_url[0] == '/'
-                or next_url.startswith(settings.DOMAIN)
-                or next_url.startswith(settings.CAS_SERVER_URL)
-                or next_url.startswith(settings.MFR_SERVER_URL)):
-            raise HTTPError(http.InvalidURL)
+    :param service_url: the service url to check
+    :param kwargs: None
+    :return: True if valid, False otherwise
+    """
 
-    if auth.logged_in:
-        if not log_out:
-            if next_url:
-                return redirect(next_url)
-            return redirect('dashboard')
-        # redirect user to CAS for logout, return here w/o authentication
-        return auth_logout(redirect_url=request.url)
+    if not (service_url[0] == '/' or
+            service_url.startswith(settings.DOMAIN) or
+            service_url.startswith(settings.CAS_SERVER_URL) or
+            # TODO: why mfr domain, how about api and waterbutler
```

Comment:
MFR is our secret auth provider, don't tell anyone.

But actually, unlike WaterButler or the API (aside from the Browsable API, which uses a different mechanism), the MFR can be viewable on its own, and may require login.